Issue 6476 - Fix build failure with GCC 15

mreynolds389 opened the pull request:
Description:

Most of the failures are our use of function pointer with a generic typedef with unknown parameters (e.g. IFP)

There are other IFP we use, but they are not triggering build failures yet.

I embraced using "caddr_t", because converting "caddr_t" to "void *" requires an even bigger change to the existing code, and caddr_t (although "old") is actually quite flexible and works nicely with our existing code.

Relates: https://github.com/389ds/389-ds-base/issues/6476

tbordaz commented on ldap/servers/slapd/back-ldbm/misc.c:
```diff
@@ -329,7 +329,7 @@ ldbm_txn_ruv_modify_context(Slapi_PBlock *pb, modify_context *mc)
     struct backentry *bentry;
     entry_address bentry_addr;
-    IFP fn = NULL;
+    int32_t (*fn)(Slapi_PBlock *, char *, Slapi_Mods *) = NULL;
```
IIUC fn = replica_ruv_smods_for_op(Slapi_PBlock *pb, char **uniqueid, Slapi_Mods **smods).
 should not we still call it with &uniqueid and &smods

mreynolds389 replied:
> IIUC fn = replica_ruv_smods_for_op(Slapi_PBlock *pb, char **uniqueid, Slapi_Mods **smods). should not we still call it with &uniqueid and &smods

Indeed.  I think the compiler missed this because pblock.c was not updated for it.  And there was a failure in replication acceptance test.  Ok testing new fix to see if it passes...

tbordaz reviewed (approved):
Large to review. A minor question.
LGTM

droideck commented on ldap/servers/plugins/syntaxes/syntax_common.c:
```diff
@@ -16,7 +16,7 @@ int
     struct mr_plugin_def mr_plugin_table[],
     size_t mr_plugin_table_size,
-    IFP matching_rule_plugin_init)
+    int32 (*matching_rule_plugin_init)(Slapi_PBlock *))
```
`int32_t` typo

Also, `matching_rule_plugin_init` is defined as `int` everywhere, but I don't have a hard opinion on its consistency.

mreynolds389 replied:
> `int32_t` typo
> 
> Also, `matching_rule_plugin_init` is defined as `int` everywhere, but I don't have a hard opinion on its consistency.

Yeah trying to not to make this patch bigger than it needs to be.  To be consistent I made all function pointer declarations use int32_t, but to update all the calling functions would be a lot of work.

droideck commented on ldap/servers/plugins/collation/orfilter.c:
```diff
@@ -901,7 +901,7 @@ or_filter_index(Slapi_PBlock *pb)
+    int32_t (*mrINDEX_FN)(Slapi_PBlock *) = NULL;
     auto struct berval **mrVALUES = NULL;
     auto char *mrOID = NULL;
     auto int mrQUERY_OPERATOR;
```
While we are here, should we remove `auto` in other places too? It's useless in modern C, as far as I know.

mreynolds389 replied:
> While we are here, should we remove `auto` in other places too? It's useless in modern C, as far as I know.

Fixed

droideck reviewed (approved):
A couple of minor issues; otherwise - looks good!

mreynolds389 committed:
Issue 6476 - Fix build failure with GCC 15

Description:

Most of the failures are our use of function pointer with a generic
typedef with unknown parameters (e.g. IFP)

There are other IFP we use, but they are not triggering build failures
yet.

Relates: https://github.com/389ds/389-ds-base/issues/6476

Reviewed by: tbordaz & spichugi (Thanks!!)
